src/profile/hooks/useProfileMutations.ts

```diff
@@ -0,0 +1,145 @@
+>;
+
+// cache is updated with cache.modify({fields...})
+// where properties of the "fields" object are fields to update
+// the format is: {"property": <function returning new value>}
+// example: fields:{ phones:()=>listOfPhones, primaryPhone:()=>newPrimaryPhone}
+// this helper function creates those objects from listed "fields"
+// and returns same field from given data source "updatedProfile"
+function updateFieldCreator(
+  fields: UpdatableProfileDataKeys[],
+  updatedProfile: UpdateProfileData
+): CacheUpdateProps {
+  return fields.reduce((currentValue, nextField) => {
+    currentValue[nextField] = () => updatedProfile[nextField];
+    return currentValue;
+  }, {} as CacheUpdateProps);
+}
+
+function getCacheUpdateFields(
+  dataType: EditDataType,
+  updatedProfile: UpdateProfileData
+): CacheUpdateProps {
+  const fields: UpdatableProfileDataKeys[] = [];
+  if (dataType === 'phones') {
+    fields.push(dataType, 'primaryPhone');
+  } else if (dataType === 'emails') {
+    fields.push(dataType, 'primaryEmail');
+  } else if (dataType === 'addresses') {
+    fields.push(dataType, 'primaryAddress');
+  } else if (dataType === basicDataType) {
+    fields.push('firstName', 'nickname', 'lastName');
+  } else if (dataType === additionalInformationType) {
+    fields.push('language');
+  }
+  return updateFieldCreator(fields, updatedProfile);
+}
+
+function updateCache(
+  cache: ApolloCache<UpdateProfileRoot>,
+  result: UpdateResult,
+  dataType: EditDataType,
+  profileData: ProfileRoot | null | undefined,
+  updateProfileData: ProfileContextData['updateProfileData']
+) {
+  const updatedProfile: UpdateProfileData | null | undefined =
+    result.data?.updateMyProfile?.profile;
+  const identity =
+    profileData &&
+    cache.identify((profileData.myProfile as unknown) as Reference);
+  if (!updatedProfile || !identity) {
+    throw new Error('Unable to update cache2');
+  }
+  cache.modify({
+    id: identity,
+    fields: getCacheUpdateFields(dataType, updatedProfile) as Modifiers,
+  });
+  const newData = cache.readQuery({
+    query: MY_PROFILE,
+  });
+  updateProfileData(newData as ProfileRoot);
+}
+
+export function useProfileMutations({
+  dataType,
+}: {
+  dataType: EditDataType;
+}): MutationReturnType {
+  const { data: profileData, updateProfileData } = useContext(ProfileContext);
+  const [updateProfile, { error, loading }] = useMutation<
+    UpdateProfileRoot,
+    UpdateMyProfileVariables
+  >(UPDATE_PROFILE, {
+    update: (cache, result) =>
+      updateCache(cache, result, dataType, profileData, updateProfileData),
+  });
+
+  const update: UpdateProfileFunction = (formValues, profile) => {
+    const currentProfile = profile || profileData;
+    if (!currentProfile) {
+      return Promise.reject({ error: 'No profile data' });
+    }
+    const variables = updateMutationVariables(formValues, currentProfile);
+    return updateProfile({ variables });
+  };
+
+  return { update, error, loading, profileData };
```

Comment:
Did not find usages for error and loading, maybe there's no need to return those?

Reply:
Except in useProfileMutations.test.ts. UI does not use them because Promise.reject() is catched instead. But in tests error and loading are directly tested. Should they be kept there?

Reply:
Are there plans on using them in the UI in future? If not, it might simplify things to delete them if they are not needed.

Reply:
Ok, I removed them and changed the test just to  check error wont alter profile data